javalib/src/main/scala/java/util/LinkedHashMap.scala

```diff
@@ -12,72 +12,150 @@
 
-class LinkedHashMap[K, V] private (inner: mutable.LinkedHashMap[Box[K], V],
-    accessOrder: Boolean) extends HashMap[K, V](inner) {
+class LinkedHashMap[K, V](initialCapacity: Int, loadFactor: Float,
+    accessOrder: Boolean)
+    extends HashMap[K, V](initialCapacity, loadFactor) {
   self =>
 
-  def this() =
-    this(mutable.LinkedHashMap.empty[Box[K], V], false)
-
-  def this(initialCapacity: Int, loadFactor: Float, accessOrder: Boolean) = {
-    this(mutable.LinkedHashMap.empty[Box[K], V], accessOrder)
-    if (initialCapacity < 0)
-      throw new IllegalArgumentException("initialCapacity < 0")
-    else if (loadFactor < 0.0)
-      throw new IllegalArgumentException("loadFactor <= 0.0")
-  }
+  import LinkedHashMap._
+
+  /** Node that was least recently created (or accessed under access-order). */
+  private var eldest: Node[K, V] = _
+
+  /** Node that was most recently created (or accessed under access-order). */
+  private var youngest: Node[K, V] = _
 
   def this(initialCapacity: Int, loadFactor: Float) =
     this(initialCapacity, loadFactor, false)
 
   def this(initialCapacity: Int) =
-    this(initialCapacity, LinkedHashMap.DEFAULT_LOAD_FACTOR)
+    this(initialCapacity, HashMap.DEFAULT_LOAD_FACTOR)
+
+  def this() =
+    this(HashMap.DEFAULT_INITIAL_CAPACITY)
 
   def this(m: Map[_ <: K, _ <: V]) = {
-    this()
+    this(m.size())
     putAll(m)
   }
 
-  override def get(key: scala.Any): V = {
-    val value = super.get(key)
+  private def asMyNode(node: HashMap.Node[K, V]): Node[K, V] =
+    node.asInstanceOf[Node[K, V]]
+
+  private[util] override def newNode(key: K, hash: Int, value: V,
+      previous: HashMap.Node[K, V],
+      next: HashMap.Node[K, V]): HashMap.Node[K, V] = {
+    new Node(key, hash, value, previous, next, null, null)
+  }
+
+  private[util] override def nodeWasAccessed(node: HashMap.Node[K, V]): Unit = {
     if (accessOrder) {
-      val boxedKey = Box(key.asInstanceOf[K])
-      if (value != null || containsKey(boxedKey)) {
-        inner.remove(boxedKey)
-        inner(boxedKey) = value
+      val myNode = asMyNode(node)
+      if (myNode.younger ne null) {
+        removeFromOrderedList(myNode)
+        appendToOrderedList(myNode)
       }
     }
-    value
   }
 
-  override def put(key: K, value: V): V = {
-    val oldValue = {
-      if (accessOrder) {
-        val old = remove(key)
-        super.put(key, value)
-        old
-      } else {
-        super.put(key, value)
-      }
-    }
-    val iter = entrySet().iterator()
-    if (iter.hasNext && removeEldestEntry(iter.next()))
-      iter.remove()
-    oldValue
+  private[util] override def nodeWasAdded(node: HashMap.Node[K, V]): Unit = {
+    appendToOrderedList(asMyNode(node))
+    if (removeEldestEntry(eldest))
+      removeNode(eldest)
+  }
+
+  private[util] override def nodeWasRemoved(node: HashMap.Node[K, V]): Unit =
+    removeFromOrderedList(asMyNode(node))
+
+  private def appendToOrderedList(node: Node[K, V]): Unit = {
```

Comment:
`prepend`?

Reply:
No, this really is an append. Newer nodes appear *later* in iteration order, so they appear in the append side of the list (at the end).